gprofiler/merge.py

```diff
@@ -92,33 +176,39 @@ def parse_perf_script(script: str):
             match = SAMPLE_REGEX.match(sample)
             if match is None:
                 raise Exception("Failed to match sample")
-            yield match.groupdict()
+            sample_dict = match.groupdict()
+
+            pid = int(sample_dict["pid"])
+            process_name = sample_dict["comm"]
+            stack = sample_dict["stack"]
+            if stack is not None:
+                pid_to_collapsed_stacks_counters[pid][collapse_stack(stack, process_name)] += 1
+            pid_to_name.setdefault(pid, process_name)
         except Exception:
             logger.exception(f"Error processing sample: {sample}")
+    return pid_to_collapsed_stacks_counters, pid_to_name
 
 
-def merge_perfs(perf_all: Iterable[Mapping[str, str]], process_perfs: Mapping[int, Mapping[str, int]]) -> str:
+def merge_perfs(
+    system_perf_pid_to_stacks_counter: ProcessToStackSampleCounters,
+    pid_to_name: ProcessIdToNameMapping,
+    process_perfs: ProcessToStackSampleCounters,
+) -> str:
     per_process_samples: MutableMapping[int, int] = Counter()
-    new_samples: MutableMapping[str, int] = Counter()
-    process_names = {}
-    for parsed in perf_all:
-        try:
-            pid = int(parsed["pid"])
-            if pid in process_perfs:
-                per_process_samples[pid] += 1
-                process_names[pid] = parsed["comm"]
```

Comment:
I just realized there was a bug here (and it's also relevant for the new code): the `comm` may be thread-based, and until https://github.com/Granulate/gprofiler/pull/69 we always saw `comm` of threads. Now we should be seeing the `comm` of the process, but in older kernels, we'll still see the threads' one. So this code will set the `comm` of the last thread in the `perf script` output, for *all* other threads in that process...

Reply:
Creating an issue for it